core/drivers/crypto/hisilicon/sec_authenc.c

```diff
@@ -0,0 +1,824 @@
+
+	memcpy(ae_drv_ctx->civ + NONCE_OFFSET, dinit->nonce.data,
+	       dinit->nonce.length);
+	memcpy(ae_drv_ctx->aiv + NONCE_OFFSET, dinit->nonce.data,
+	       dinit->nonce.length);
+
+	ae_drv_ctx->aiv[0] = flags;
+	ae_drv_ctx->aiv[IV_LAST_BYTE1] = data_size & IV_LAST_BYTE_MASK;
+	data_size >>= IV_BYTE_OFFSET;
+	ae_drv_ctx->aiv[IV_LAST_BYTE2] = data_size & IV_LAST_BYTE_MASK;
+	data_size >>= IV_BYTE_OFFSET;
+	ae_drv_ctx->aiv[IV_LAST_BYTE3] = data_size & IV_LAST_BYTE_MASK;
+
+	ae_drv_ctx->civ[0] = cl;
+	ae_drv_ctx->civ[MAX_IV_SIZE - 1] = IV_CTR_INIT;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_aead_get_dma(struct authenc_ctx *ae_drv_ctx)
+{
+	ae_drv_ctx->key_dma = virt_to_phys(ae_drv_ctx->key);
+	if (!ae_drv_ctx->key_dma) {
+		EMSG("Failed to map key_dma");
+		return TEE_ERROR_GENERIC;
+	}
+
+	ae_drv_ctx->civ_dma = virt_to_phys(ae_drv_ctx->civ);
+	if (!ae_drv_ctx->civ_dma) {
+		EMSG("Failed to map civ_dma");
+		return TEE_ERROR_GENERIC;
+	}
+
+	ae_drv_ctx->tag_dma = virt_to_phys(ae_drv_ctx->tag);
+	if (!ae_drv_ctx->tag_dma) {
+		EMSG("Failed to map tag_dma");
+		return TEE_ERROR_GENERIC;
+	}
+
+	ae_drv_ctx->src_dma = virt_to_phys(ae_drv_ctx->src.data);
+	ae_drv_ctx->dst_dma = virt_to_phys(ae_drv_ctx->dst.data);
+
+	if (ae_drv_ctx->mode == C_MODE_GCM)
+		return TEE_SUCCESS;
+	ae_drv_ctx->aiv_dma = virt_to_phys(ae_drv_ctx->aiv);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_aead_data_alloc(struct authenc_ctx *ae_drv_ctx)
+{
+	ae_drv_ctx->src.length = ae_drv_ctx->payload_len +
+				 ae_drv_ctx->aad.length;
+	ae_drv_ctx->src.data = malloc(ae_drv_ctx->src.length);
+	if (!ae_drv_ctx->src.data) {
+		EMSG("src is NULL");
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	ae_drv_ctx->dst.length = ae_drv_ctx->src.length;
+	ae_drv_ctx->dst.data = malloc(ae_drv_ctx->dst.length);
+	if (!ae_drv_ctx->dst.data) {
+		EMSG("dst is NULL");
+		free(ae_drv_ctx->src.data);
+		ae_drv_ctx->src.data = NULL;
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result ae_soft_calc_init(struct authenc_ctx *ae_drv_ctx,
+				    struct drvcrypt_authenc_init *dinit)
+{
+	struct crypto_authenc_ctx *ae_soft_ctx = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (ae_drv_ctx->algo == C_ALG_AES && ae_drv_ctx->mode == C_MODE_GCM) {
+		ae_drv_ctx->is_hw_support = false;
+		ae_soft_ctx = ae_drv_ctx->ae_soft_ctx;
+		ret = ae_soft_ctx->ops->init(ae_soft_ctx,
+			(TEE_OperationMode)(dinit->encrypt == 0),
+			dinit->key.data, dinit->key.length,
+			dinit->nonce.data, dinit->nonce.length,
+			dinit->tag_len, dinit->aad_len,
+			dinit->payload_len);
+		if (ret)
+			EMSG("Fail to init by soft ctx");
+
+		return ret;
+	}
+
+	return TEE_ERROR_NOT_IMPLEMENTED;
+}
+
+static TEE_Result sec_authenc_initialize(struct drvcrypt_authenc_init *dinit)
+{
+	struct authenc_ctx *ae_drv_ctx = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+	bool hw_support = false;
+
+	ret = authenc_init_params_check(dinit);
+	if (ret)
+		return ret;
+
+	ae_drv_ctx = (struct authenc_ctx *)dinit->ctx;
```

Comment:
There's no need to cast a void pointer, if there are other places like this please fix those too.